### ISSUE TITLE
cache: cargo, only caching installed artifacts

### ISSUE THREAD

**Issue by behnam** [travis-ci/travis-ci#8396]
In Rust ecosystem, `cargo` performs two major tasks: one is project-dependent build and package management, and another is user-level registry caching and binary installation.

The `cache: cargo` option in Travis config file is described here: https://docs.travis-ci.com/user/caching/#Rust-Cargo-cache

This config enables caching for artifacts from both these functionalities, which is not always desired. There are many cases that the second functionality is desired---which is basically caching downloaded source files and built binaries for external files, downloaded from the crates.io registry---but no caching is desired for the first set of functionalities---which is build artifacts for the source code in current repository.

So, basically, there's need for an option to enable caching for `$HOME/.cargo`, but NOT enable it for `$TRAVIS_BUILD_DIR/target`.

An example of this is in `rust-lang/cargo` project itself, where we need to cache the binary installation of `mdbook` pacakge, but not caching `cargo`'s build files. Basically, with caching `$HOME/.cargo`, this line will be executed much faster on every run:

https://github.com/rust-lang/cargo/blob/36ddeff03a31f710095cfbe49287503bb60e294d/.travis.yml#L24-L25

Is it currently possible?

**Comment by BanzaiMan** [comment]
Use directories: https://docs.travis-ci.com/user/caching#Arbitrary-directories

```yaml
cache:
  directories:
    - target
```

**Comment by behnam** [comment]
Uh, I missed that feature! Thanks, @BanzaiMan!

Although, I think what we want in this case is:
```yml
cache:
  directories:
    - $HOME/.cargo
```

**Comment by BanzaiMan** [comment]
Oops. Sorry about the confusion there!
